website/blog/deployers.md

```diff
@@ -0,0 +1,610 @@
+</div>
+
+### Telemetry
+
+Finally, we implement the telemetry methods. All logs produced by a weavelet are
+received by the `HandleLogEntry` method. Our deployer uses a pretty printer from
+Service Weaver's `logging` library to print the logs to stdout. Similarly, all
+traces produced by a weavelet are received by the `HandleTraceSpans` function.
+For simplicity, our deployer ignore traces.
+
+```golang
+// Responsibility 3: Telemetry.
+func (h *handler) HandleLogEntry(_ context.Context, entry *protos.LogEntry) error {
+    pp := logging.NewPrettyPrinter(colors.Enabled())
+    fmt.Println(pp.Format(entry))
+    return nil
+}
+
+func (h *handler) HandleTraceSpans(context.Context, []trace.ReadOnlySpan) error {
+    return nil
+}
+```
+
+<div class="note">
+<code>weaver multi</code> writes logs and traces to files. <code>weaver gke</code> exports logs and traces to <a href="https://cloud.google.com/logging">Cloud Logging</a> and <a href="https://cloud.google.com/trace">Cloud Trace</a>.
+</div>
+
+### Main
+
+Finally, we implement `main`. We create a deployer, spawn the main component,
+and block.
+
+```go
+func main() {
+    flag.Parse()
+    d := &deployer{weavelets: map[string]*weavelet{}}
+    d.spawn("main")
+    select {} // block forever
+}
+```
+
+## Advanced Deployer Features
+
+The multiprocess deployer in the previous section was designed to be as simple
+as possible. Real-world deployers, on the other hand, require a number of more
+advanced features. Enumerating and explaining how to implement these features is
+beyond the scope of this blog post, but we'll summarize some advanced features
+here. You can also review the implementations of our [`weaver
+multi`][weaver_github] and [`weaver gke`][weaver_gke_github] deployers as
+reference.
+
+- **Failure Detection**. A deployer should detect when a weavelet has failed and
+  notify the other weavelets accordingly. You can use the `Envelope.GetHealth`
+  method to check the health of a weavelet.
+
+- **Routing**. A deployer should support [routed components][routing] by
+  monitoring the load on routed components and generating [routing
+  assignments][Assignment] that balance this load. You can use the
+  `Envelope.GetLoad` method to get the load of a weavelet.
+
+- **Rollouts**. A deployer should implement [versioned rollouts][versioning],
+  allowing one version of an application to be rolled out as a replacement to a
+  previously running version.
+
+- **Tooling**. A deployer should provide tooling to inspect and debug the state
+  of an application. `weaver multi status`, `weaver multi dashboard`, and
+  `weaver multi logs`, for example, can be used to inspect applications deployed
+  with `weaver multi deploy`.
+
```

Comment:
A few things are unclear to me at this point:
* Is deployer a single binary? If so, the diagram in the beginning of the post is a bit misleading, since we say that envelopes interact with weavelets using UNIX pipes.
* Is deployer a long running process/service? Given that deployer has to react to weavelet crashes, load changes, rollouts etc., my mental model of deployer at this point is that of something similar to Borg Master.

Lastly, I feel that the work done by a deployer is much more than just deployment of binaries. In that sense, the term deployer seems a bit narrow. Do we expect deployer to take on even more responsibilities in the future?

Reply:
Although this is a guide on deployer implementation, I think giving out some information about the following will help the user:
* Where does the deployer run?
* What happens if a deployer crashes? Is it supposed to maintain state continuity after a restart?

Reply:
Thanks Nipun! This is exactly the kind of feedback I was hoping for!

> Is deployer a single binary? If so, the diagram in the beginning of the post is a bit misleading, since we say that envelopes interact with weavelets using UNIX pipes.

Ah, I think it wasn't clear from our deployer where subprocesses were getting launched. I updated the description to explain this more carefully. The deployer above was indeed a single process. It runs every weavelet in its own subprocess though. This is what I was trying to convey with the figure. 

> Is deployer a long running process/service? Given that deployer has to react to weavelet crashes, load changes, rollouts etc., my mental model of deployer at this point is that of something similar to Borg Master.

Yes, a real-world deployer would likely be long running. I added a new bullet point here explaining that and gave our GKE deployer as an example.

> Lastly, I feel that the work done by a deployer is much more than just deployment of binaries. In that sense, the term deployer seems a bit narrow. Do we expect deployer to take on even more responsibilities in the future?

Good point. We'll brainstorm a better term.

> Where does the deployer run?

I hope the GKE example I added helps explain better where a deployer may run.

> What happens if a deployer crashes? Is it supposed to maintain state continuity after a restart?

Ditto.